src/tests/fuzzing/fuzz_pkcs15_decode.c

```diff
@@ -20,60 +20,92 @@
     int (* decode_entries[])(struct sc_pkcs15_card *, struct sc_pkcs15_object *,
             const u8 **nbuf, size_t *nbufsize) = {
         sc_pkcs15_decode_prkdf_entry, sc_pkcs15_decode_pukdf_entry,
         sc_pkcs15_decode_skdf_entry, sc_pkcs15_decode_cdf_entry,
         sc_pkcs15_decode_dodf_entry, sc_pkcs15_decode_aodf_entry
     };
     int algorithms[] = { SC_ALGORITHM_RSA, SC_ALGORITHM_EC, SC_ALGORITHM_GOSTR3410, SC_ALGORITHM_EDDSA };
-    size_t i;
 
+    /* Establish context for fuzz app*/
+    sc_establish_context(&ctx, "fuzz");
     if (!ctx)
-        sc_establish_context(&ctx, "fuzz");
-    if (!p15card) {
-        card.ctx = ctx;
-        p15card = sc_pkcs15_card_new();
-        if (p15card) {
-            p15card->card = &card;
-        }
+        return 0;
+
+    /* Erase possible readers from ctx */
+    while (list_size(&ctx->readers)) {
+        sc_reader_t *rdr = (sc_reader_t *) list_get_at(&ctx->readers, 0);
+        _sc_delete_reader(ctx, rdr);
+    }
+    if (ctx->reader_driver->ops->finish != NULL)
+        ctx->reader_driver->ops->finish(ctx);
+
+    /* Create virtual reader */
+    ctx->reader_driver = sc_get_fuzz_driver();
+    fuzz_add_reader(ctx, Data, Size);
+    reader = sc_ctx_get_reader(ctx, 0);
+
+    /* Connect card to reader */
+    if (sc_connect_card(reader, &card)) {
+        sc_release_context(ctx);
+        return 0;
     }
 
     for (i = 0; i < 4; i++) {
         struct sc_pkcs15_object *obj;
-        const u8 *p = Data;
-        size_t len = Size;
-        obj = calloc(1, sizeof *obj);
-        while (SC_SUCCESS == decode_entries[i](p15card, obj, &p, &len)) {
+        const u8 *p = NULL;
+        size_t len = 0;
+        if (!(obj = calloc(1, sizeof *obj)))
+            goto err;
+        fuzz_get_chunk(reader, &buf, &buf_len);
+        p = buf;
+        len = (size_t) buf_len;
+        while (SC_SUCCESS == decode_entries[i](p15card, obj, &p, (size_t *) &len)) {
             sc_pkcs15_free_object(obj);
-            obj = calloc(1, sizeof *obj);
+            if (!(obj = calloc(1, sizeof *obj)))
+                goto err;
         }
         sc_pkcs15_free_object(obj);
     }
 
+    fuzz_get_chunk(reader, &buf, &buf_len);
     for (i = 0; i < sizeof algorithms/sizeof(int); i++) {
         struct sc_pkcs15_pubkey *pubkey = calloc(1, sizeof *pubkey);
         pubkey->algorithm = algorithms[i];
-        sc_pkcs15_decode_pubkey(ctx, pubkey, Data, Size);
+        sc_pkcs15_decode_pubkey(ctx, pubkey, buf, buf_len);
         sc_pkcs15_free_pubkey(pubkey);
     }
 
-    struct sc_pkcs15_tokeninfo *tokeninfo = sc_pkcs15_tokeninfo_new();
-    sc_pkcs15_parse_tokeninfo(ctx, tokeninfo, Data, Size);
+    fuzz_get_chunk(reader, &buf, &buf_len);
+    tokeninfo = sc_pkcs15_tokeninfo_new();
+    sc_pkcs15_parse_tokeninfo(ctx, tokeninfo, buf, buf_len);
     sc_pkcs15_free_tokeninfo(tokeninfo);
 
-    sc_pkcs15_parse_unusedspace(Data, Size, p15card);
+    sc_pkcs15_parse_unusedspace(buf, buf_len, p15card);
```

Comment:
I think this should use a new `buf` and `buf_len` gathered from `fuzz_get_chunk()`. This is aliased with the data parsed by `sc_pkcs15_parse_tokeninfo()`.

Reply:
Indeed, thank you.